### ISSUE TITLE
getAuthToken doesn't check expiration

### ISSUE THREAD

**Issue by mattkrick** [SocketCluster/socketcluster-server#1]
Currently, middleware that uses `getAuthToken` has to manually check for it's expiration. Middleware is an authorization process & the errors it throws should be authorization errors. So, how do we handle an authentication error, eg expired JWT?

I propose that on every middleware handler (eg `verifyInboundEvent, verifyOutboundEvent`) we check the expiration of the JWT. If it's expired, we ignore the error (if any) thrown by the middleware & throw an `Authentication Error` (we could reuse `_processTokenError`).

The only edge case is if the if the middleware doesn't care if the token is expired. Since this is likely rare, we could make this the second argument in the `next()` call, eg `next('Not authorized', true)`. This is probably cleaner than having them store a static & accessing it after the call.

The value in this is that we can digest this message internally to know if the socket has lost authentication. The value for devs is that they don't have to parse a string, rather an error type.

**Comment by jondubois** [comment]
@mattkrick I like your thinking.

When you say 'the errors it throws', I assume you mean on both the server and the client? (I.e. emitted as an 'authenticationFail' or  'authenticationError' or 'authFail'... (not sure which convention we want to use)... I guess this is related to the discussion in https://github.com/SocketCluster/socketcluster-client/pull/28

I also agree that we should put the client-side error inside an Error object, maybe with a ```type``` property which could tell us exactly which middleware this error comes from?

Maybe we should create a new issue/RFC to plan for SC v4 (from the main SocketCluster/socketcluster repo) and link all these sub-issues from it (including this one). I think we should have all the major changes documented and discussed in GitHub (using issues) before we start to implement because it's much easier to fix flaws in the logic before they're implemented in code.

**Comment by jondubois** [comment]
I opened a new issue here: https://github.com/SocketCluster/socketcluster/issues/137

**Comment by jondubois** [comment]
@mattkrick I think the 'authorizationError' and 'authorizationError' error should be triggered in a more explicit way - Maybe the developer should be able to choose what is most appropriate (instead of letting SC decide).

The reason for this is that we don't want to rely too much on the token... The JWT token is not necessarily the only way that the user could be authenticated... Maybe they want to use a one-time password (instead of a JWT token) - Maybe the user doesn't have to be logged in to perform a specific action. I think it's important to be explicit.

What if the user could specify what kind of error to emit to the client when they call the ```next()``` callback?

E.g.
```next('authorizationError', 'You are not allowed to access this method')``` could trigger an authorizationError event while ```next('authenticationError', 'You are not authenticated')``` would trigger an authenticationError on the client.

**Comment by mattkrick** [comment]
Ahhhhh, didn't know socketcluster offered auth aside from JWTs, neat!

My concern is this: a JWT is dangerous without expiration. Checking for expiration currently must be performed on every middleware. If I have 1 middleware for each table in my database, then if I forget once, boom, I'm open to attacks (plus it's not very dry to rewrite it over & over).

Could we write an internal `JWT-check-exp` middleware that is turned on by a server option? (or better, on by default?)

That JWT middleware would basically be:
```
token = getAuthToken()
if token.exp < Date.now()
tokenError = new Error('Your JWT has expired')
tokenError.type = 'Token Error'; 
next(tokenError);
```
Then, if they want to ignore expirations or use something else, they just turn off the flag, and everyone else gets a nice, secure connection. 

Regarding error types in the `next` call, I'm totally onboard. Instead of 2 params, we could also have them pass in an Error object & then we dehydrate it behind the scenes before we emit it. OR emit a pure object that we serialize.

Now the question is, how do they listen to specific errors? I suggest from the server we emit a `#middlewareError`. Upon receiving it on the client, we rehydrate the error & emit it on a local `middlewareError` that they can listen for. Alternatively, if we emitted a serialized object from the server, then we could skip the preprocessing step. This would also allow them to send more than `type` and `message` (although type should be required & throw if not included).

**Comment by mattkrick** [comment]
On second thought, we have a chance to homogenize all the middleware and really clean up the API.
Currently, each one has different args.
Let's offer two: a `request` object and a `next` function.
The request object's properties are populated based on the type of middleware.
For example, subscription has a `socket` and `channel` prop.
If there's an error, we add that error to the request. This should be familiar to anyone who's used express.
Then, `next()` is pure, without argument.
This also allows us to pass in extra props to the next middleware (eg If i receive a subscription, I can parse out the userId).

**Comment by jondubois** [comment]
@mattkrick I fully agree about making middleware API consistency. I agree about rehydrating all errors on the client so that they will always be Error objects. I also agree with allowing the developer to specify whether or not their channel subscription should wait for authentication to take place (maybe an extra argument/option passed to the ```socket.subscribe(...)``` call.

I kind of like the idea of a ```jwt-expiration``` middleware which sits in front of all other middlewares and would trigger a client-side ```authTokenExpired``` when it detects that a token has expired. But I'm thinking that this middleware should not block the flow (it should always flow on to the next middleware even if the token is expired and let the next middleware decide whether or not to block the request/event).

Following the same line of logic, I'm also backtracking on the initial idea to have the middleware ```next()``` call support two kinds of errors ```next('authorizationError', '...')``` and ```next('authenticationError', '...')```.

I think we should separate authentication from authorization further and it seems to me that middleware itself has nothing to do with 'authentication' and everything to do with 'authorization'.

I think that if a socket is unauthenticated, that doesn't necessarily mean that it's also unauthorized to perform a specific action*. The other side of this argument is also true; if a socket is authenticated, that doesn't necessarily mean that they are authorized to perform a specific action.

So basically, I agree with adding a jwt-expiration middleware, but that middleware would never block a request - All it would do would be check the token's expiration and emit an ```authTokenExpired``` event on the client-side (or similar) if it is expired - Also it would remove the token form the socket on the backend... So calling ```socket.getAuthToken()``` would now return null... THEN the following middleware (e.g. MIDDLEWARE_PUBLISH, MIDDLEWARE_SUBSCRIBE, ...) can decide whether or not to block the request based on the absence of the token.

* I think it will be quiet common for people to want to have a Reddit-style workflow where anyone can read/subscribe to a channel, but only people who are logged in with an account can post/publish.

**Comment by jondubois** [comment]
I kind of like passing the error to the ```next('error string')``` in the middleware though, I find it more explicit than adding an error property to the request object.

**Comment by mattkrick** [comment]
I like it! Socket cluster handles the authentication part and passes it to
the dev middleware and they can choose what to do with it.
I checked out some other middleware APIs yesterday, redux has a clean one
and it passes a single action to next. I like this a little better because
it allows the dev to create a brand new object, they arent confined to
modifying the request object. Thoughts?

**Comment by jondubois** [comment]
@mattkrick Cool. I updated the list of planned changes here: https://github.com/SocketCluster/socketcluster/issues/137

Pretty much all of them are based on your ideas.
I've been thinking through different scenarios and these changes seems to fit nicely but feel free to suggest more stuff in case anything else can be improved or if I misunderstood any of your use cases.

Thanks for all the suggestions and pull requests!
